src/deserializer.rs

```diff
@@ -0,0 +1,172 @@
+use std::{convert};
+use std::io::{Cursor, Read};
+
+use byteorder::{self, ReadBytesExt};
+
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum DeserializationError {
+    UnexpectedTag,
+    ShortData,
+    ExtraData,
+    IntegerOverflow,
+    InvalidValue,
+}
+
+impl convert::From<byteorder::Error> for DeserializationError {
+    fn from(e: byteorder::Error) -> DeserializationError {
+        return match e {
+            byteorder::Error::UnexpectedEOF => DeserializationError::ShortData,
+            _ => panic!("Unexpected error!"),
+        }
+    }
+}
+
+pub type DeserializationResult<T> = Result<T, DeserializationError>;
+
+pub struct Deserializer {
+    reader: Cursor<Vec<u8>>,
+}
+
+impl Deserializer {
+    pub fn new(data: Vec<u8>) -> Deserializer {
+        return Deserializer{
+            reader: Cursor::new(data),
+        }
+    }
+
+    fn _read_length(&mut self) -> DeserializationResult<usize> {
+        let b = try!(self.reader.read_u8());
+        // TODO: handle lengths greater than 128
+        assert!(b & 0x80 == 0);
+        return Ok((b & 0x7f) as usize);
+    }
+
+    fn _read_with_tag<T, F>(&mut self, expected_tag: u8, body: F) -> DeserializationResult<T>
+            where F: Fn(Vec<u8>) -> DeserializationResult<T> {
+        let tag = try!(self.reader.read_u8());
+        // TODO: only some of the bits in the first byte are for the tag
+        if tag != expected_tag {
+            return Err(DeserializationError::UnexpectedTag);
+        }
+        let length = try!(self._read_length());
+        let mut data = vec![0; length];
+        let n = self.reader.read(&mut data).unwrap();
+        if n != length {
+            return Err(DeserializationError::ShortData);
+        }
+        return body(data);
+    }
+
+    pub fn finish(self) -> DeserializationResult<()> {
+        if self.reader.position() as usize != self.reader.get_ref().len() {
+            return Err(DeserializationError::ExtraData);
+        }
+        return Ok(());
+    }
+
+    pub fn read_bool(&mut self) -> DeserializationResult<bool> {
+        return self._read_with_tag(1, |data| {
+            if data == b"\x00" {
+                return Ok(false);
+            } else if data == b"\xff" {
+                return Ok(true)
+            } else {
+                return Err(DeserializationError::InvalidValue);
+            }
+        });
+    }
+
+    pub fn read_int(&mut self) -> DeserializationResult<i64> {
+        return self._read_with_tag(2, |data| {
+            if data.len() > 8 {
```

Comment:
Is there a rust bignum for when you support larger integers? I guess that doesn't matter for this PR.